### ISSUE TITLE
files uploaded to server are broken after 2.0.0

### ISSUE THREAD

**Issue by jessewee** [salim-lachdhaf/ftpConnect#77]
it works fine when i switch to 1.0.1

**Comment by salim-lachdhaf** [comment]
please groupe your issues in single request and past an example of your code to analyse the issue.

**Comment by jessewee** [comment]
[test_ftp.zip](https://github.com/salim-lachdhaf/ftpConnect/files/9650978/test_ftp.zip)
[TestFtp.zip](https://github.com/salim-lachdhaf/ftpConnect/files/9650979/TestFtp.zip)

test_ftp is a flutter project, it's the client app using ftpConnect.
TestFtp is a Android project, it's the server app using apache ftpserver.

1. file uploaded to server are broken after 2.0.0
    change the ftpConnect version to 2.0.2
    run the project
    select a picture and upload
    check the picture from server device, it can not be opened
    change the ftpConnect version to 1.0.1 and retry, the picture in server device is fine

2. call disconnect when uploading, the upload continues
    change the ftpConnect version to 1.0.1
    pick a large file
    start uploading
    after the progress reached about 3%, call disconnect, the uploading continues

**Comment by jessewee** [comment]
3.while uploading, pull off the power line of server device, it takes minutes to receive the disconnect result.
    here's the log

2022-09-27 11:31:29.962 15072-15072 AutofillManager         com.example.test_ftp                 V  requestHideFillUi(null): anchor = null
2022-09-27 11:31:29.985 15072-15823 BLASTBufferQueue        com.example.test_ftp                 E  [SurfaceView[com.example.test_ftp/com.example.test_ftp.MainActivity]#3](f:0,a:2) isEGL=1, mPendingRelease.size()=1, mMaxAcquiredBuffers=4, currentMaxAcquiredBufferCount=2
2022-09-27 11:31:30.021 15072-15823 BLASTBufferQueue        com.example.test_ftp                 E  [SurfaceView[com.example.test_ftp/com.example.test_ftp.MainActivity]#3](f:0,a:4) isEGL=1, mPendingRelease.size()=2, mMaxAcquiredBuffers=4, currentMaxAcquiredBufferCount=2
2022-09-27 11:31:30.069 15072-16014 flutter                 com.example.test_ftp                 I  [27.09.2022 11:31:30.069] Connecting...
2022-09-27 11:31:30.128 15072-16014 flutter                 com.example.test_ftp                 I  [27.09.2022 11:31:30.128] > USER admin
2022-09-27 11:31:30.950 15072-16014 flutter                 com.example.test_ftp                 I  [27.09.2022 11:31:30.950] < 220 Service ready for new user.
                                                                                                    331 User name okay, need password for admin.
2022-09-27 11:31:30.953 15072-16014 flutter                 com.example.test_ftp                 I  [27.09.2022 11:31:30.953] > PASS 123456
2022-09-27 11:31:31.760 15072-16014 flutter                 com.example.test_ftp                 I  [27.09.2022 11:31:31.759] < 230 User logged in, proceed.
2022-09-27 11:31:31.762 15072-16014 flutter                 com.example.test_ftp                 I  [27.09.2022 11:31:31.762] Connected!
2022-09-27 11:31:31.786 15072-16014 flutter                 com.example.test_ftp                 I  [27.09.2022 11:31:31.785] Upload File: /storage/emulated/0/图片和视频/jellyfish-160-mbps-4k-uhd-h264.mkv
2022-09-27 11:31:31.789 15072-16014 flutter                 com.example.test_ftp                 I  [27.09.2022 11:31:31.789] Entering transfer mode TransferMode.binary
2022-09-27 11:31:31.791 15072-16014 flutter                 com.example.test_ftp                 I  [27.09.2022 11:31:31.791] > TYPE I
2022-09-27 11:31:32.600 15072-16014 flutter                 com.example.test_ftp                 I  [27.09.2022 11:31:32.599] < 200 Command TYPE okay.
2022-09-27 11:31:32.603 15072-16014 flutter                 com.example.test_ftp                 I  [27.09.2022 11:31:32.603] > EPSV
2022-09-27 11:31:33.414 15072-16014 flutter                 com.example.test_ftp                 I  [27.09.2022 11:31:33.413] < 229 Entering Passive Mode (|||36601|)
2022-09-27 11:31:33.481 15072-16014 flutter                 com.example.test_ftp                 I  [27.09.2022 11:31:33.481] > STOR jellyfish-160-mbps-4k-uhd-h264.mkv
2022-09-27 11:31:33.485 15072-16014 flutter                 com.example.test_ftp                 I  [27.09.2022 11:31:33.485] Opening DataSocket to Port 36601
2022-09-27 11:31:34.024 15072-16014 flutter                 com.example.test_ftp                 I  [27.09.2022 11:31:34.024] < 150 File status okay; about to open data connection.
2022-09-27 11:31:34.025 15072-16014 flutter                 com.example.test_ftp                 I  [27.09.2022 11:31:34.025] Start uploading...
2022-09-27 11:34:25.326 15072-15072 ZoomWindowDecorView     com.example.test_ftp                 V  setLastReportedMergedConfiguration mZoomDisplayHeight: 1080 getDecorView.195302329
2022-09-27 11:34:25.330 15072-15872 BufferQueueProducer     com.example.test_ftp                 I  [ViewRootImpl[com.example.test_ftp/com.example.test_ftp.MainActivity]#2(BLAST Consumer)2](id:3ae000000002,api:1,p:15072,c:15072) disconnect: api 1
2022-09-27 11:34:25.330 15072-15872 BLASTBufferQueue        com.example.test_ftp                 E  BLASTBufferItemConsumer::onDisconnect()
2022-09-27 11:34:25.330 15072-15872 BufferQueueProducer     com.example.test_ftp                 I  [ViewRootImpl[com.example.test_ftp/com.example.test_ftp.MainActivity]#2(BLAST Consumer)2](id:3ae000000002,api:1,p:15072,c:15072) connect: api=1 producerControlledByApp=true
2022-09-27 11:34:25.337 15072-16015 BufferQueueProducer     com.example.test_ftp                 I  [SurfaceView[com.example.test_ftp/com.example.test_ftp.MainActivity]#3(BLAST Consumer)3](id:3ae000000003,api:1,p:15072,c:15072) disconnect: api 1
2022-09-27 11:34:25.341 15072-16015 BLASTBufferQueue        com.example.test_ftp                 E  BLASTBufferItemConsumer::onDisconnect()
2022-09-27 11:34:25.341 15072-16015 BufferQueueProducer     com.example.test_ftp                 I  [SurfaceView[com.example.test_ftp/com.example.test_ftp.MainActivity]#3(BLAST Consumer)3](id:3ae000000003,api:1,p:15072,c:15072) connect: api=1 producerControlledByApp=true
2022-09-27 11:34:25.368 15072-16822 BLASTBufferQueue        com.example.test_ftp                 E  [ViewRootImpl[com.example.test_ftp/com.example.test_ftp.MainActivity]#2](f:0,a:1) isEGL=1, mPendingRelease.size()=1, mMaxAcquiredBuffers=4, currentMaxAcquiredBufferCount=2
2022-09-27 11:34:25.369 15072-15872 BufferQueueProducer     com.example.test_ftp                 I  [ViewRootImpl[com.example.test_ftp/com.example.test_ftp.MainActivity]#2(BLAST Consumer)2](id:3ae000000002,api:1,p:15072,c:15072) disconnect: api 1
2022-09-27 11:34:25.369 15072-15872 BLASTBufferQueue        com.example.test_ftp                 E  BLASTBufferItemConsumer::onDisconnect()
2022-09-27 11:34:25.369 15072-15872 BufferQueueProducer     com.example.test_ftp                 I  [ViewRootImpl[com.example.test_ftp/com.example.test_ftp.MainActivity]#2(BLAST Consumer)2](id:3ae000000002,api:1,p:15072,c:15072) connect: api=1 producerControlledByApp=true
2022-09-27 11:34:25.383 15072-16822 BLASTBufferQueue        com.example.test_ftp                 E  [ViewRootImpl[com.example.test_ftp/com.example.test_ftp.MainActivity]#2](f:0,a:1) isEGL=1, mPendingRelease.size()=2, mMaxAcquiredBuffers=4, currentMaxAcquiredBufferCount=2
2022-09-27 11:34:27.766 15072-15072 ZoomWindowDecorView     com.example.test_ftp                 V  setLastReportedMergedConfiguration mZoomDisplayHeight: 1080 getDecorView.195302329
2022-09-27 11:34:27.779 15072-15872 BufferQueueProducer     com.example.test_ftp                 I  [ViewRootImpl[com.example.test_ftp/com.example.test_ftp.MainActivity]#2(BLAST Consumer)2](id:3ae000000002,api:1,p:15072,c:15072) disconnect: api 1
2022-09-27 11:34:27.779 15072-15872 BLASTBufferQueue        com.example.test_ftp                 E  BLASTBufferItemConsumer::onDisconnect()
2022-09-27 11:34:27.780 15072-15872 BufferQueueProducer     com.example.test_ftp                 I  [ViewRootImpl[com.example.test_ftp/com.example.test_ftp.MainActivity]#2(BLAST Consumer)2](id:3ae000000002,api:1,p:15072,c:15072) connect: api=1 producerControlledByApp=true
2022-09-27 11:34:28.349 15072-15072 ZoomWindowDecorView     com.example.test_ftp                 V  setLastReportedMergedConfiguration mZoomDisplayHeight: 2400 getDecorView.195302329
2022-09-27 11:34:28.350 15072-15872 BufferQueueProducer     com.example.test_ftp                 I  [ViewRootImpl[com.example.test_ftp/com.example.test_ftp.MainActivity]#2(BLAST Consumer)2](id:3ae000000002,api:1,p:15072,c:15072) disconnect: api 1
2022-09-27 11:34:28.350 15072-15872 BLASTBufferQueue        com.example.test_ftp                 E  BLASTBufferItemConsumer::onDisconnect()
2022-09-27 11:34:28.351 15072-15872 BufferQueueProducer     com.example.test_ftp                 I  [ViewRootImpl[com.example.test_ftp/com.example.test_ftp.MainActivity]#2(BLAST Consumer)2](id:3ae000000002,api:1,p:15072,c:15072) connect: api=1 producerControlledByApp=true
2022-09-27 11:34:28.353 15072-16015 BufferQueueProducer     com.example.test_ftp                 I  [SurfaceView[com.example.test_ftp/com.example.test_ftp.MainActivity]#3(BLAST Consumer)3](id:3ae000000003,api:1,p:15072,c:15072) disconnect: api 1
2022-09-27 11:34:28.353 15072-16015 BLASTBufferQueue        com.example.test_ftp                 E  BLASTBufferItemConsumer::onDisconnect()
2022-09-27 11:34:28.353 15072-16015 BufferQueueProducer     com.example.test_ftp                 I  [SurfaceView[com.example.test_ftp/com.example.test_ftp.MainActivity]#3(BLAST Consumer)3](id:3ae000000003,api:1,p:15072,c:15072) connect: api=1 producerControlledByApp=true
2022-09-27 11:34:28.370 15072-15872 BufferQueueProducer     com.example.test_ftp                 I  [ViewRootImpl[com.example.test_ftp/com.example.test_ftp.MainActivity]#2(BLAST Consumer)2](id:3ae000000002,api:1,p:15072,c:15072) disconnect: api 1
2022-09-27 11:34:28.370 15072-15872 BLASTBufferQueue        com.example.test_ftp                 E  BLASTBufferItemConsumer::onDisconnect()
2022-09-27 11:34:28.370 15072-15872 BufferQueueProducer     com.example.test_ftp                 I  [ViewRootImpl[com.example.test_ftp/com.example.test_ftp.MainActivity]#2(BLAST Consumer)2](id:3ae000000002,api:1,p:15072,c:15072) connect: api=1 producerControlledByApp=true
2022-09-27 11:35:11.130 15072-16015 BufferQueueProducer     com.example.test_ftp                 I  [SurfaceView[com.example.test_ftp/com.example.test_ftp.MainActivity]#3(BLAST Consumer)3](id:3ae000000003,api:1,p:15072,c:15072) disconnect: api 1
2022-09-27 11:35:11.130 15072-16015 BLASTBufferQueue        com.example.test_ftp                 E  BLASTBufferItemConsumer::onDisconnect()
2022-09-27 11:35:11.130 15072-15072 BufferQueueProducer     com.example.test_ftp                 I  [SurfaceView[com.example.test_ftp/com.example.test_ftp.MainActivity]#3(BLAST Consumer)3](id:3ae000000003,api:0,p:-1,c:15072) disconnect: api -1
2022-09-27 11:35:11.130 15072-15072 BLASTBufferQueue        com.example.test_ftp                 I  [SurfaceView[com.example.test_ftp/com.example.test_ftp.MainActivity]#3] destructor()
2022-09-27 11:35:11.130 15072-15072 BufferQueueConsumer     com.example.test_ftp                 I  [SurfaceView[com.example.test_ftp/com.example.test_ftp.MainActivity]#3(BLAST Consumer)3](id:3ae000000003,api:0,p:-1,c:15072) disconnect
2022-09-27 11:35:11.134 15072-15872 BufferQueueProducer     com.example.test_ftp                 I  [ViewRootImpl[com.example.test_ftp/com.example.test_ftp.MainActivity]#2(BLAST Consumer)2](id:3ae000000002,api:1,p:15072,c:15072) disconnect: api 1
2022-09-27 11:35:11.134 15072-15872 BLASTBufferQueue        com.example.test_ftp                 E  BLASTBufferItemConsumer::onDisconnect()
2022-09-27 11:35:11.136 15072-16822 BLASTBufferQueue        com.example.test_ftp                 I  releaseBufferCallbackThunk bufferId:64733747085334 framenumber:76 blastBufferQueue is dead
2022-09-27 11:35:11.144 15072-15072 BLASTBufferQueue        com.example.test_ftp                 I  [ViewRootImpl[com.example.test_ftp/com.example.test_ftp.MainActivity]#2] destructor()
2022-09-27 11:35:11.144 15072-15072 BufferQueueConsumer     com.example.test_ftp                 I  [ViewRootImpl[com.example.test_ftp/com.example.test_ftp.MainActivity]#2(BLAST Consumer)2](id:3ae000000002,api:0,p:-1,c:15072) disconnect
2022-09-27 11:35:11.168 15072-16822 BLASTBufferQueue        com.example.test_ftp                 I  releaseBufferCallbackThunk bufferId:64733747085333 framenumber:6 blastBufferQueue is dead
2022-09-27 11:35:14.064 15072-15072 BufferQueueConsumer     com.example.test_ftp                 I  [](id:3ae000000004,api:0,p:-1,c:15072) connect: controlledByApp=false
2022-09-27 11:35:14.064 15072-15072 BLASTBufferQueue        com.example.test_ftp                 I  [ViewRootImpl[com.example.test_ftp/com.example.test_ftp.MainActivity]#4] constructor()
2022-09-27 11:35:14.065 15072-15872 BufferQueueProducer     com.example.test_ftp                 I  [ViewRootImpl[com.example.test_ftp/com.example.test_ftp.MainActivity]#4(BLAST Consumer)4](id:3ae000000004,api:1,p:15072,c:15072) connect: api=1 producerControlledByApp=true
2022-09-27 11:35:14.070 15072-15072 BufferQueueConsumer     com.example.test_ftp                 I  [](id:3ae000000005,api:0,p:-1,c:15072) connect: controlledByApp=false
2022-09-27 11:35:14.070 15072-15072 BLASTBufferQueue        com.example.test_ftp                 I  [SurfaceView[com.example.test_ftp/com.example.test_ftp.MainActivity]#5] constructor()
2022-09-27 11:35:14.072 15072-16015 BufferQueueProducer     com.example.test_ftp                 I  [SurfaceView[com.example.test_ftp/com.example.test_ftp.MainActivity]#5(BLAST Consumer)5](id:3ae000000005,api:1,p:15072,c:15072) connect: api=1 producerControlledByApp=true
2022-09-27 11:35:14.086 15072-16822 SurfaceComposerClient   com.example.test_ftp                 D  VRR [FRTC] client handle [bufferId:18446744073709551615 framenumber:0] [ffffffff, ffffffff]
2022-09-27 11:35:14.097 15072-15823 SurfaceComposerClient   com.example.test_ftp                 D  VRR [FRTC] client handle [bufferId:18446744073709551615 framenumber:0] [ffffffff, ffffffff]
2022-09-27 11:36:52.215 15072-16014 flutter                 com.example.test_ftp                 I  SocketException: Connection reset by peer (OS Error: Connection reset by peer, errno = 104), address = 192.168.43.75, port = 43096
2022-09-27 11:36:52.218 15072-16014 flutter                 com.example.test_ftp                 I  [27.09.2022 11:36:52.217] Disconnecting...
2022-09-27 11:36:52.218 15072-16014 flutter                 com.example.test_ftp                 I  [27.09.2022 11:36:52.218] > QUIT
2022-09-27 11:36:52.290 15072-15823 BLASTBufferQueue        com.example.test_ftp                 E  [SurfaceView[com.example.test_ftp/com.example.test_ftp.MainActivity]#5](f:0,a:2) isEGL=1, mPendingRelease.size()=1, mMaxAcquiredBuffers=4, currentMaxAcquiredBufferCount=2
2022-09-27 11:37:02.232 15072-16014 flutter                 com.example.test_ftp                 I  [27.09.2022 11:37:02.231] Disconnected!

**Comment by jessewee** [comment]
what' the principle of retry?
when uploading i restarted the server and the client just failed.

                                                                                                                                                                         thanks

**Comment by TakaTeke** [comment]
I have the same problem.
Uploading Image File, PDF File, Video File to the host server corrupts the file.
Even if I try specifying the transferType to binary, it still breaks.
By the way, the Json file is not broken.

**Comment by TakaTeke** [comment]
I tested with a PDF file. The file size before and after uploading is different. Probably because the file got corrupted. It seems that it is because the TransferType is set to ascii and the binary file is uploaded.
    Local file: 306,988 bytes
    Host file: 304,287 bytes
Note: I added "onProgress:" to the "_ftpConnect.uploadFile" instruction and checked the transmission status, but the result of "onProgress:" is displayed as 306,988 bytes. But the actual size of the file on the host side is 304,287 bytes.

When I added "_ftpConnect.sendCustomCommand('TYPE I');" to "_uploadStepByStep" and executed it, the response was "FTPReply = [code= 200, message= 200 Type set to I.]" and it was successful. The file size is now also:
    Local file: 306,988 bytes
    Host file: 306,988 bytes

Also, I added "_ftpConnect.transferType = TransferType.binary;" and executed it, but I get "FTPReply = [code= 150, message= 150 ASCII data connection for XXXXXXXX.pdf.]". It did not become "Type set to I". I don't know why. Maybe it's in the wrong place as you describe it.

I saw your source code "ftp_socket.dart", it defaults to "TransferType.auto" and sends commands with 'TYPE A'. Perhaps you need to identify the type of file you are sending, such as "TYPE A" for text-based files and "Type I" for binary mode files?

As a temporary measure, when uploading files in binary mode, adding "_ftpConnect.sendCustomCommand('TYPE I');" to "_uploadStepByStep" as follows works fine.

<<Execute statement>>
Future<void> _uploadStepByStep(String fileId) async {
  _ftpConnect =
      FTPConnect(ftpHost, user: ftpUser, pass: ftpPass, debug: true);
  try {
    await _log('Connecting to FTP ...');
    await _ftpConnect.connect();
    await _ftpConnect.changeDirectory('$ftpHomeDirectory/pdf');
    final file = File('$pdfDirectory/$fileId');
    await _ftpConnect.sendCustomCommand('TYPE I');
    await _log('Uploading ...');
    _ftpConnect.uploadFile(file, onProgress: (percent, received, fileSize) {
      print('percent: $percent / received: $received / fileSize: $fileSize');
    }).then((value) async {
      await _log('file uploaded successfully');
      await _ftpConnect.disconnect();
    });
  } catch (e) {
    await _log('Error: ${e.toString()}');
  }
}


<<Execution result log>>
flutter: Connecting to FTP ...
flutter: [20.10.2022 00:06:37.630] Connecting...
flutter: [20.10.2022 00:06:37.648] Connection established, waiting for welcome message...
flutter: [20.10.2022 00:06:37.953] < FTPReply =  [code= 220, message= 220 netftpXX XXXXXXXXXXXXXXXXXXXXX Ready.]
flutter: [20.10.2022 00:06:37.953] > USER XXXXXXXXXX
flutter: [20.10.2022 00:06:38.255] < FTPReply =  [code= 331, message= 331 Password required for XXXXXXXXXX.]
flutter: [20.10.2022 00:06:38.256] > PASS XXXXXXXXXXXX
flutter: [20.10.2022 00:06:38.558] < FTPReply =  [code= 230, message= 230 User XXXXXXXXXX logged in.]
flutter: [20.10.2022 00:06:38.558] Connected!
flutter: [20.10.2022 00:06:38.561] > CWD homepage/pdf
flutter: [20.10.2022 00:06:38.863] < FTPReply =  [code= 250, message= 250 CWD command succesful.]
flutter: [20.10.2022 00:06:38.864] > TYPE I
flutter: [20.10.2022 00:06:39.165] < FTPReply =  [code= 200, message= 200 Type set to I.]
flutter: Uploading ...
flutter: [20.10.2022 00:06:40.168] Upload File: /Users/taka/testFile2/pdf/XXXXXXXXXX.pdf
flutter: [20.10.2022 00:06:40.174] > PASV
flutter: [20.10.2022 00:06:40.477] < FTPReply =  [code= 227, message= 227 Entering Passive Mode (XXX,XXX,XX,XXX,XXX,XX)]
flutter: [20.10.2022 00:06:40.477] > STOR XXXXXXXXXX.pdf
flutter: [20.10.2022 00:06:40.478] Opening DataSocket to Port XXXXX
flutter: [20.10.2022 00:06:40.798] < FTPReply =  [code= 150, message= 150 BINARY data connection for XXXXXXXXXX.pdf.]
flutter: [20.10.2022 00:06:40.798] Start uploading...
flutter: percent: 21.35 / received: 65536 / fileSize: 306988
flutter: percent: 42.7 / received: 131072 / fileSize: 306988
flutter: percent: 64.04 / received: 196608 / fileSize: 306988
flutter: percent: 85.39 / received: 262144 / fileSize: 306988
flutter: percent: 100.0 / received: 306988 / fileSize: 306988
flutter: [20.10.2022 00:06:41.142] < FTPReply =  [code= 226, message= 226 STOR "XXXXXXXXXX.pdf" success.]
flutter: [20.10.2022 00:06:41.142] File Uploaded!
flutter: file uploaded successfully
flutter: [20.10.2022 00:06:42.145] Disconnecting...
flutter: [20.10.2022 00:06:42.145] > QUIT
flutter: [20.10.2022 00:06:42.448] < FTPReply =  [code= 221, message= 221 Goodbye.]
flutter: [20.10.2022 00:06:42.448] Disconnected!

**Comment by artem312** [comment]
@salim-lachdhaf i fixed this issue, please check pull request

**Comment by Ghali01** [comment]
> @salim-lachdhaf i fixed this issue, please check pull request

Did you fix the download also?

**Comment by salim-lachdhaf** [comment]
@jessewee could you test the this version on master branch if it correct your issue before release the new version ?

**Comment by artem312** [comment]
> > @salim-lachdhaf i fixed this issue, please check pull request
> 
> Did you fix the download also?

As i understand it fixed transfer type in common

**Comment by jessewee** [comment]
> @jessewee could you test the this version on master branch if it correct your issue before release the new version ?

sorry, i'm not familiar with github, and don't login often.
And i can't the version you said. the master branch is still on version 2.0.2 to me.

**Comment by crifurch** [comment]
> > @jessewee could you test the this version on master branch if it correct your issue before release the new version ?
> 
> sorry, i'm not familiar with github, and don't login often. And i can't find the version you said. the master branch is still on version 2.0.2 to me.

Yes now in master version 2.0.2 with fix before bump version to 2.0.3 and make release pls 
1 replace ftp_connect: ^2.0.2  with
```
ftp_connect:
    git:
      url: https://github.com/salim-lachdhaf/ftpConnect.git
      ref: master 
```
2 do pub get
3 try reproduce problem
4 if problem was solved for you write about it here

**Comment by salim-lachdhaf** [comment]
version 2.0.3 was released.
